### PR TITLE
test(e2e): `TestBitcoinDepositRefund`, `WithdrawBitcoinMultipleTimes`

### PR DESCRIPTION
**Pull request by swift1337**
# Description

This PR adds E2E tests for `TestBitcoinDepositRefund` and `WithdrawBitcoinMultipleTimes`. It also introduces several improvements to testing suite:
- [x] Support for testify's `assert` and `require` packages in e2e testing
- [x] Support for debugging e2e test using Delve debugger `./e2e/scripts/debug.sh $test $args`
- [x] Shortcut for running a single e2e test `./e2e/scripts/run.sh $test $args`
- [x] Support for `utils.WaitCctxByInboundHash()` with some handy filters

Closes: https://github.com/zeta-chain/node/issues/1469

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [x] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: e2e/e2etests/e2etests.go [hunk 1]
```diff
@@ -56,6 +56,7 @@ const (
 	 Test transfer of Bitcoin asset across chains
 	*/
 	TestBitcoinDepositName                = "bitcoin_deposit"
+	TestBitcoinDepositRefund              = "bitcoin_deposit_refund"
```
Comment (lumtis) [inline e2e/e2etests/e2etests.go:59]:
```suggestion
	TestBitcoinDepositRefundName              = "bitcoin_deposit_refund"
```

File: e2e/e2etests/e2etests.go [hunk 2]
```diff
@@ -332,6 +333,13 @@ var AllE2ETests = []runner.E2ETest{
 		},
 		TestBitcoinDeposit,
 	),
+	runner.NewE2ETest(
+		TestBitcoinDepositRefund,
+		"deposit Bitcoin into ZEVM; expect refund", []runner.ArgDefinition{
+			{Description: "amount in btc", DefaultValue: "0.001"},
+		},
+		TestDepositBTCRefund,
```
Comment (lumtis) [inline e2e/e2etests/e2etests.go:341]:
```suggestion
		TestBitcoinDepositRefund,
```

File: e2e/e2etests/e2etests.go [hunk 3]
```diff
@@ -332,6 +333,13 @@ var AllE2ETests = []runner.E2ETest{
 		},
 		TestBitcoinDeposit,
 	),
+	runner.NewE2ETest(
+		TestBitcoinDepositRefund,
```
Comment (lumtis) [inline e2e/e2etests/e2etests.go:337]:
```suggestion
		TestBitcoinDepositRefundName,
```

File: e2e/e2etests/test_bitcoin_deposit_refund.go [hunk 4]
```diff
@@ -1,56 +1,41 @@
...
-//}
+import (
+	"strconv"
+
+	"github.com/stretchr/testify/require"
+	"github.com/zeta-chain/zetacore/e2e/runner"
+	"github.com/zeta-chain/zetacore/e2e/utils"
+)
+
+func TestDepositBTCRefund(r *runner.E2ERunner, args []string) {
```
Comment (lumtis) [inline e2e/e2etests/test_bitcoin_deposit_refund.go:11]:
```suggestion
func TestBitcoinDepositRefund(r *runner.E2ERunner, args []string) {
```

**Review by lumtis — COMMENTED** [review]
We'll need to add the new test in `cmd/zetae2e/local/local.go` to execute in CI

File: e2e/e2etests/test_bitcoin_deposit_refund.go [hunk 5]
```diff
@@ -1,56 +1,74 @@
...
+	// ARRANGE
+	// Given BTC address
+	r.SetBtcAddress(r.Name, false)
+
+	// Given "Live" BTC network
+	stop := r.MineBlocksIfLocalBitcoin()
+	defer stop()
+
+	// Given amount to send
+	require.Len(r, args, 1)
```
Comment (gartnera) [inline e2e/e2etests/test_bitcoin_deposit_refund.go:26]: If we're going to start using `require`, we should just skip straight to [`assert`](https://pkg.go.dev/github.com/stretchr/testify/assert) style where you do `a := assert.New(t)` then `a.Len(args, 1)` so you don't have to provide `t` every time.

Technically, `assert` calls `Fail` instead of `FailNow` but we can just have `Fail` stop immediately in the e2e tests.

File: e2e/e2etests/helper_bitcoin.go [hunk 6]
```diff
@@ -55,50 +55,41 @@ func withdrawBTCZRC20(r *runner.E2ERunner, to btcutil.Address, amount *big.Int)
...
 	cctx := utils.WaitCctxMinedByInboundHash(r.Ctx, receipt.TxHash.Hex(), r.CctxClient, r.Logger, r.CctxTimeout)
-	if cctx.CctxStatus.Status != crosschaintypes.CctxStatus_OutboundMined {
-		panic(fmt.Errorf("cctx status is not OutboundMined"))
-	}
+	require.Equal(
+		r,
+		crosschaintypes.CctxStatus_OutboundMined,
+		cctx.CctxStatus.Status,
+		"cctx status is not OutboundMined",
+	)
```
Comment (lumtis) [inline e2e/e2etests/helper_bitcoin.go:84]: Same as for the receipt. I think we could introduce some helpers directly checking for some status state

File: e2e/e2etests/test_bitcoin_deposit_refund.go [hunk 7]
```diff
@@ -1,56 +1,74 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/zeta-chain/zetacore/e2e/runner"
+	"github.com/zeta-chain/zetacore/e2e/utils"
+	"github.com/zeta-chain/zetacore/x/crosschain/types"
+	zetabitcoin "github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin"
+)
+
+func TestBitcoinDepositRefund(r *runner.E2ERunner, args []string) {
+	// ARRANGE
```
Comment (lumtis) [inline e2e/e2etests/test_bitcoin_deposit_refund.go:17]: I didn't know this comment convention, looks good, we might want to extend it to all tests. I think we can open an issue for this.

File: e2e/e2etests/test_bitcoin_deposit_refund.go [hunk 8]
```diff
@@ -1,56 +1,74 @@
...
+	refundTx, err := r.BtcRPCClient.GetTransaction(btcTxHash)
+	require.NoError(r, err, refundTx)
+
+	// Finally, check the refund transaction details
+	refundTxDetails := refundTx.Details[0]
+	assert.Equal(r, "receive", refundTxDetails.Category)
+	assert.Equal(r, r.BTCDeployerAddress.EncodeAddress(), refundTxDetails.Address)
+	assert.NotEmpty(r, refundTxDetails.Amount)
+
+	r.Logger.InfoLoud("Sent %f BTC to TSS with invalid memo, got refund of %f BTC", amount, refundTxDetails.Amount)
```
Comment (lumtis) [inline e2e/e2etests/test_bitcoin_deposit_refund.go:67]:
```suggestion
	r.Logger.Info("Sent %f BTC to TSS with invalid memo, got refund of %f BTC", amount, refundTxDetails.Amount)
```
I think loud is not necessary

File: e2e/e2etests/test_bitcoin_withdraw_multiple.go [hunk 9]
```diff
@@ -1,96 +1,52 @@
...
+	"strconv"
+
+	"github.com/btcsuite/btcutil"
+	"github.com/stretchr/testify/require"
+
+	"github.com/zeta-chain/zetacore/e2e/runner"
+	"github.com/zeta-chain/zetacore/pkg/chains"
+)
+
+const defaultReceiver = "mxpYha3UJKUgSwsAz2qYRqaDSwAkKZ3YEY"
```
Comment (lumtis) [inline e2e/e2etests/test_bitcoin_withdraw_multiple.go:14]: Need comment why using this value

File: e2e/e2etests/test_bitcoin_deposit_refund.go [hunk 10]
```diff
@@ -1,56 +1,74 @@
...
+
+	// Query the BTC network to check the refund transaction
+	refundTx, err := r.BtcRPCClient.GetTransaction(btcTxHash)
+	require.NoError(r, err, refundTx)
+
+	// Finally, check the refund transaction details
+	refundTxDetails := refundTx.Details[0]
+	assert.Equal(r, "receive", refundTxDetails.Category)
+	assert.Equal(r, r.BTCDeployerAddress.EncodeAddress(), refundTxDetails.Address)
+	assert.NotEmpty(r, refundTxDetails.Amount)
```
Comment (lumtis) [inline e2e/e2etests/test_bitcoin_deposit_refund.go:65]: Why specifically using assert here?

File: e2e/e2etests/test_bitcoin_deposit_refund.go [hunk 11]
```diff
@@ -1,56 +1,74 @@
...
+
+	r.Logger.InfoLoud("Sent %f BTC to TSS with invalid memo, got refund of %f BTC", amount, refundTxDetails.Amount)
+}
+
+func parseFloat(t require.TestingT, s string) float64 {
```
Comment (lumtis) [inline e2e/e2etests/test_bitcoin_deposit_refund.go:70]: I think we should group these parse methods somewhere

File: e2e/e2etests/test_bitcoin_deposit_refund.go (hunk 5 above)
Comment (swift1337) [reply e2e/e2etests/test_bitcoin_deposit_refund.go:26]: Indeed, in this case we want to halt the execution asap. Usually I'd use assert with unit / integration tests

File: e2e/README.md [hunk 12]
```diff
@@ -1,24 +1,33 @@
 # `e2e`
 
-`e2e` is a comprehensive suite of E2E tests designed to validate the integration and functionality of the ZetaChain network, particularly its interactions with Bitcoin and EVM (Ethereum Virtual Machine) networks. This tool is essential for ensuring the robustness and reliability of ZetaChain's cross-chain functionalities.
+`e2e` is a comprehensive suite of E2E tests designed to validate the integration and functionality of the ZetaChain
```
[the remark this answers is not in the document]
Comment (swift1337) [reply e2e/README.md:3]: Yes, this is an accidental auto-formatting of Goland `(opt+cmd+L)` 😅 . Should I keep this or do a revert?

File: e2e/e2etests/test_bitcoin_deposit_refund.go (hunk 5 above)
Comment (gartnera) [reply e2e/e2etests/test_bitcoin_deposit_refund.go:26]: I didn't notice there is actually a `require.New(t)` as well which I think we should use. So:

```
r := require.New(t)
r.Len(args, 1)
```

File: e2e/e2etests/test_bitcoin_deposit_refund.go (hunk 7 above)
Comment (swift1337) [reply e2e/e2etests/test_bitcoin_deposit_refund.go:17]: It's inspired by Behavior-Driven Development (BDD) and [Cucumber](https://cucumber.io/). AAA: `Arrange-Act-Assert` or `Given-When-Then`

File: e2e/e2etests/test_bitcoin_deposit_refund.go (hunk 10 above)
Comment (swift1337) [reply e2e/e2etests/test_bitcoin_deposit_refund.go:65]: In this case, we'd want to know what is possibly wrong with any of `refundTxDetails`'s fields. For example, maybe the `refundTxDetails.Address` is valid, but `amount` and `category` aren't. In this case, we'd see the whole picture. Otherwise, you might run this test, fix one thing, and then discover another issue.

tl;dr this helps inspect test failures faster.

File: e2e/e2etests/test_bitcoin_deposit_refund.go (hunk 5 above)
Comment (swift1337) [reply e2e/e2etests/test_bitcoin_deposit_refund.go:26]: I think it's a subjective preference 😅  For me it's clearer with verbose `require.*`, `assert.*` because it visually helps to distinguish the logic from test assertions. I'll keep this as is.

**Comment by lumtis** [comment]
Need to run
```
make generate
```
Note: don't be surprised, this command takes 2 min because we run formatter in the entire codebase (we might want to find some optimization on this)

**Review by lumtis — APPROVED** [review]
👍

File: e2e/runner/runner.go [hunk 13]
```diff
@@ -295,3 +293,14 @@ func (runner *E2ERunner) PrintContractAddresses() {
 	runner.Logger.Print("ERC20:      %s", runner.ERC20Addr.Hex())
 	runner.Logger.Print("TestDappEVM:       %s", runner.EvmTestDAppAddr.Hex())
 }
+
+// Errorf logs an error message. Mimics the behavior of testing.T.Errorf
+func (runner *E2ERunner) Errorf(format string, args ...any) {
```
[the remark this answers is not in the document]
Comment (swift1337) [reply e2e/runner/runner.go:298]: They are used to support testify's `assert` and `require`

File: e2e/runner/bitcoin.go [hunk 14]
```diff
@@ -261,13 +263,9 @@ func (runner *E2ERunner) SendToTSSFromDeployerWithMemo(
 	}
 
 	stx, signed, err := btcRPC.SignRawTransactionWithWallet2(tx, inputsForSign)
-	if err != nil {
```
[the remark this answers is not in the document]
Comment (swift1337) [reply e2e/runner/bitcoin.go:264]: I'll tackle panics right after merging this one

File: go.mod [hunk 15]
```diff
@@ -347,6 +347,8 @@ replace (
 	github.com/confio/ics23/go => github.com/cosmos/cosmos-sdk/ics23/go v0.8.0
 
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+
+	github.com/rjeczalik/notify => github.com/rjeczalik/notify v0.9.3
```
[the remark this answers is not in the document]
Comment (swift1337) [reply go.mod:351]: https://github.com/zeta-chain/node/pull/2349/commits/55405551de0dcc8c8ebae81be7f83ddcd3840b75

File: e2e/e2etests/helpers.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/zeta-chain/zetacore/pkg/chains"
+	crosschaintypes "github.com/zeta-chain/zetacore/x/crosschain/types"
+)
+
+func withdrawBTCZRC20(r *runner.E2ERunner, to btcutil.Address, amount *big.Int) *btcjson.TxRawResult {
+	tx, err := r.BTCZRC20.Approve(
+		r.ZEVMAuth,
+		r.BTCZRC20Addr,
+		big.NewInt(amount.Int64()*2),
+	) // approve more to cover withdraw fee
```
[the remark this answers is not in the document]
Comment (swift1337) [reply e2e/e2etests/helpers.go:25]: I didn't change this, it was copied from another file w/o alteration

File: e2e/scripts/run.sh [hunk 17]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (swift1337) [reply e2e/scripts/run.sh:1]: It's just an alias:
```
./e2e/scripts/run.sh bitcoin_withdraw_restricted 0.001
```
vs
```
go run ./cmd/zetae2e/ run bitcoin_withdraw_restricted:0.001 --config cmd/zetae2e/config/local.yml --verbose
```

File: e2e/e2etests/test_bitcoin_withdraw_multiple.go [hunk 18]
```diff
@@ -1,96 +1,32 @@
...
+
+func WithdrawBitcoinMultipleTimes(r *runner.E2ERunner, args []string) {
+	// ARRANGE
```
[the remark this answers is not in the document]
Comment (swift1337) [reply e2e/e2etests/test_bitcoin_withdraw_multiple.go:13]: https://github.com/zeta-chain/node/pull/2349#discussion_r1646524534
